internal/locCluster.ecl

```diff
@@ -0,0 +1,314 @@
+
+            ds[i]->isModified = true;
+            vector<node*> neighs = getNeighbors(ds,ds[i],eps);
+
+            if(neighs.size() >= minpts){
+                ds[i]->ifCore = true;
+                ds[i]->parent = NULL;
+                for(uint64_t n=0; n < neighs.size(); ++n){
+                    neighs[n]->isModified = true;
+                    if(neighs[n]->ifLocal){
+                        if(neighs[n]->ifCore){
+                            Union(ds,ds[i],neighs[n]);
+                        } else {
+                            if(neighs[n]->isVisited) continue;
+                            neighs[n]->isVisited = true;
+                            Union(ds,ds[i],neighs[n]);
+                        }
+                    } else {
+                        if(!neighs[n]->ifCore){
+                            if(getNeighbors(ds,neighs[n],eps).size()>=minpts){
+                                neighs[n]->ifCore = true;
+                                Union(ds,ds[i],neighs[n]);
+                            } else {
+                                if(neighs[n]->isVisited) continue;
+                                neighs[n]->isVisited = true;
+                                Union(ds,ds[i],neighs[n]);
+                            }
+                        } else {
+                            Union(ds,ds[i],neighs[n]);
+                        }
+                    }
+                }
+            }
+        }
+    }
+    // The data structure for the return layout.
+    struct retRecord{
+    uint32_t wi;
+    uint32_t id;
+    uint32_t parentId;
+    uint32_t nodeId;
+    bool if_local;
+    bool if_core;
+    };
+    //ResultStream Interface returns the resulting rows as a stream for global merge phase.
+    //Uses retRecord datastructure to store the results of local clustering
+    class ResultStream : public RtlCInterface, implements IRowStream {
+    public:
+        ResultStream(IEngineRowAllocator *_ra, IRowStream *_ds, uint64_t minpts, double eps,
+                                    unsigned long long lnode) : ra(_ra), ds(_ds), Lnode(lnode){
+            byte* p;
+            count = 0;
+            while((p=(byte*)ds->nextRow())){
+                node temp;
+                temp.wi = *((uint16_t*)p); p += sizeof(uint16_t);
+                temp.id = *((unsigned long long*)p); p += 2*sizeof(unsigned long long);
+                temp.nodeId = *((unsigned long long*)p); p += sizeof(unsigned long long);
+                p += sizeof(bool);
+                uint32_t lenFields = *((uint32_t*)p);
+                p += sizeof(uint32_t);
+                for(uint32_t i=0; i<lenFields/sizeof(float); ++i){
+                    float f = *((float*)p); p += sizeof(float);
+                    temp.fields.push_back(f);
+                }
+                temp.ifLocal = (lnode == temp.nodeId); p += sizeof(bool);
+                temp.ifCore = false;
+                dataset.push_back(temp);
+            }
```

Comment:
This loop isn't releasing the input rows.  It will need to call rtlReleaseRow(original);

Reply:
Updated

Reply:
That change is not correct.
i) the release of the row needs to be inside the loop
ii) The original pointer needs to be preserved - p has been updated to walk through the contents.
E.g., see the next variable in the global merge example.

Reply:
Thanks for the detailed comments @ghalliday The code is corrected. The commit info is kept for review this time. Please review. Thanks.